### PR TITLE
SALTO-4858 Add missing extensions transition rules change validator

### PR DESCRIPTION
**Pull request by IlayDavid95**
Jira:
- Added change validator for workflows that will prevent deployment of transition rules that depend on Jira apps not installed in the target Jira Cloud env. This will only work for the latest iteration of Workflows by Jira (AKA WorkflowV2).

---

_Additional context for reviewer_
This PR is a part of a series of changes, followed by a FixElement and a Filter, for allowing deployment without the faulty transition rules and fixing special Forge apps transition rules, respectively.

---
_Release Notes_: 
Jira:
- Prevent deployment of workflows with transition rules that are dependent on missing Jira apps. This is only for the new version of Workflows by Jira (enabled by the enableNewWorkflowAPI flag)

### CODE REVIEW

File: packages/jira-adapter/src/client/client.ts [hunk 1]
```diff
@@ -247,4 +287,60 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
...
+    const gqlResponse = await this.gqlPost({
+      url: '/gateway/api/graphql',
+      query: QUERY_INSTALLED_APPS,
+      variables: { cloudId: `ari:cloud:jira::site/${cloudId}` },
+    })
+
+    return _.get(gqlResponse, 'data.ecosystem.appInstallationsByContext.nodes', []).map(
+      (node: { app: { id: string; name: string } }) => ({
+        id: _.get(node, 'app.id').substring(
+          EXTENSION_ID_ARI_PREFIX.length,
```
Comment (shayc331) [inline packages/jira-adapter/src/client/client.ts:335]: there is any other way to do it intead of using the length? it seems fregile

File: packages/jira-adapter/src/config/config.ts [hunk 2]
```diff
@@ -55,6 +55,7 @@ type JiraDeployConfig = definitions.UserDeployConfig &
     forceDelete: boolean
     taskMaxRetries: number
     taskRetryDelay: number
+    ignoreMissingExtensions: boolean
```
Comment (shayc331) [inline packages/jira-adapter/src/config/config.ts:58]: what is the purpose of this flag? I dont see any usage

File: packages/jira-adapter/src/change_validators/workflows/empty_validator_workflow.ts [hunk 3]
```diff
@@ -63,10 +63,10 @@ const workflowV1HasEmptyValidator = (transition: TransitionV1, invalidValidators
   })
 }
 
-const workflowV2HasEmptyValidator = (transition: WorkflowTransitionV2, invalidValidators: Set<string>): void => {
+const workflowV2HasEmptyValidator = (transition: WorkflowV2Transition, invalidValidators: Set<string>): void => {
   transition.validators?.forEach(validator => {
     if (isEmptyValidatorV2(validator)) {
-      invalidValidators.add(validator.parameters.ruleType)
+      invalidValidators.add(_.get(validator, 'parameters.ruleType'))
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/workflows/empty_validator_workflow.ts:69]: why did you change it?

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+const getConditionRules: (
+  elemID: ElemID,
+  conditionGroup?: WorkflowV2ConditionGroup,
+) => TransitionRuleWithElemIDType[] = (elemID, conditionGroup) => {
+  if (conditionGroup === undefined) {
+    return []
+  }
+  let conditionRules: TransitionRuleWithElemIDType[] = []
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts:62]: nit: 
you can use const, and then push to the list

File: packages/jira-adapter/src/filters/workflow/transition_structure.ts [hunk 5]
```diff
@@ -160,15 +160,15 @@ export const walkOverTransitionIds = (transition: WorkflowTransitionV1, func: (v
...
 }
 
-export const walkOverTransitionIdsV2 = (transition: WorkflowTransitionV2, func: (value: Value) => void): void => {
+export const walkOverTransitionIdsV2 = (transition: WorkflowV2Transition, func: (value: Value) => void): void => {
   transition.actions
     ?.filter(
       action =>
-        action.parameters?.appKey === SCRIPT_RUNNER_POST_FUNCTION_TYPE &&
-        !_.isEmpty(action.parameters.scriptRunner?.transitionId),
+        _.get(action, 'parameters.appKey') === SCRIPT_RUNNER_POST_FUNCTION_TYPE &&
```
Comment (shayc331) [inline packages/jira-adapter/src/filters/workflow/transition_structure.ts:167]: I think we should improve our scheme guard intead of using `_.get`, if something goes wrong we wont know it

File: packages/jira-adapter/src/filters/workflowV2/types.ts [hunk 6]
```diff
@@ -83,15 +83,31 @@ export type WorkflowStatusAndPort = {
   port?: number
 }
 
-export type WorkflowTransitionV2 = {
+export type WorkflowV2TransitionRuleParameters = {
+  appKey?: string
+  key?: string
```
Comment (shayc331) [inline packages/jira-adapter/src/filters/workflowV2/types.ts:88]: are they can not be defined togheter?
is one of them must be defined?
maybe we can use Xor between them? in the SAAS we use `ts-xor` library for it, check it out

**Review by shayc331 — COMMENTED** [review]
great job! it is looking really good 🏆 
add some comments

File: packages/jira-adapter/src/filters/workflowV2/types.ts (hunk 6 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/filters/workflowV2/types.ts:88]: if ruleKey starts with `system` they don't have to be there. if ruleKey starts with `connect` then `appKey` must be defined and if ruleKey starts with `forge` then `key` must be defined.

Hence, they might be defined together (I don't think there is a constraint against it in Jira), they might not be defined at all, and sometimes only one of them can be defined.

File: packages/jira-adapter/src/filters/workflow/transition_structure.ts (hunk 5 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/filters/workflow/transition_structure.ts:167]: But appKey is an optional parameter that is only required in case of a `connect` type rule.

File: packages/jira-adapter/src/config/config.ts (hunk 2 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/config/config.ts:58]: This is a relic of history. At first the PR had a usage for it (used in the change validator) and then it was decided to use a fixElement instead (next PR) and I simply left this here due to being lazy I guess. The next PR (that comes out of this one) properly uses it.

File: packages/jira-adapter/src/change_validators/workflows/empty_validator_workflow.ts (hunk 3 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflows/empty_validator_workflow.ts:69]: Because ruleType is completely optional.

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+type TransitionRuleWithElemIDType = WorkflowV2TransitionRule & { elemID: ElemID }
+
+const isRuleType = (value: any): value is RuleType => Object.values(RuleType).includes(value)
+
+const getRuleTypeFromRuleKey: (ruleKey: string) => RuleType = ruleKey => {
+  const ruleType = ruleKey.split(':')[0]
+  if (!isRuleType(ruleType)) {
+    throw Error(`Unexpected rule type ${ruleType}, expected of one of ${Object.values(RuleType)}`)
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts:49]: Will change.

File: packages/jira-adapter/src/client/client.ts [hunk 8]
```diff
@@ -41,6 +42,39 @@ const DEFAULT_PAGE_SIZE: Required<definitions.ClientPageSizeConfig> = {
 
 const RATE_LIMIT_HEADER_PREFIX = 'x-ratelimit-'
 
+export const GET_CLOUD_ID_URL = '/_edge/tenant_info'
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:45]: Not redundant because it is used in tests.

File: packages/jira-adapter/src/client/client.ts [hunk 9]
```diff
@@ -165,14 +200,19 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
     url: string
     query: string
     variables?: Record<string, unknown>
+    headers?: Record<string, string>
   }): Promise<graphQLResponseType> {
+    let headers = args.headers
+    if (headers === undefined) {
+      headers = args.url === GQL_BASE_URL_GATEWAY ? EXPERIMENTAL_API_HEADERS : PRIVATE_API_HEADERS
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:207]: In the end I just changed the query so it doesn't need the headers.

File: packages/jira-adapter/src/client/client.ts (hunk 1 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:335]: Can split by "/" and take the element in position [1].

File: packages/jira-adapter/src/change_validators/workflows/empty_validator_workflow.ts (hunk 3 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflows/empty_validator_workflow.ts:69]: Parameters don't necessarily have ruleType.

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    (acc, app) => {
+      acc[app.id] = app
+      return acc
+    },
+    {} as Record<string, ExtensionType>,
+  )
+
+export const missingExtensionsTransitionRulesChangeValidator = (client: JiraClient): ChangeValidator => {
+  return async changes => {
+    const installedExtensionsMap = await getInstalledExtensionsMap(client)
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts:150]: Done.

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+export const getExtensionIdFromWorkflowTransitionRule = (transitionRule: WorkflowV2TransitionRule) => {
+  const ruleType = getRuleTypeFromWorkflowTransitionRule(transitionRule)
+  const extensionKey = getExtensionKeyFromWorkflowTransitionRule(transitionRule)
+  if (extensionKey === undefined) {
+    return undefined
+  }
+
+  switch (ruleType) {
+    case RuleType.Connect:
+      return extensionKey.split('__')[0]
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts:143]: Done

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    case RuleType.Forge:
+      return extensionKey.split('/')[1]
+    default:
+      return undefined
+  }
+}
+
+export const isValidExtensionTransitionRule = (transitionRule: WorkflowV2TransitionRule): boolean =>
+  getRuleTypeFromWorkflowTransitionRule(transitionRule) !== RuleType.System &&
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts]: Removed.

File: packages/jira-adapter/src/client/client.ts [hunk 13]
```diff
@@ -58,9 +86,79 @@ const isGraphQLResponse = createSchemeGuard<graphQLResponseType>(
...
+const UPM_INSTALLED_APPS_RESPONSE_SCHEME = Joi.object({
+  data: Joi.object({
+    plugins: Joi.array()
+      .items({
+        name: Joi.string().required(),
+        key: Joi.string().required(),
+        enabled: Joi.bool().required(),
+        userInstalled: Joi.bool().required(),
+      })
+      .required(),
```
Comment (shayc331) [inline packages/jira-adapter/src/client/client.ts:109]: I think we should add `.unkown(true)` for `plugins` and `data`

File: packages/jira-adapter/src/config/config.ts (hunk 2 above)
Comment (shayc331) [reply packages/jira-adapter/src/config/config.ts:58]: so please remove it from this pr and add it when you need it, it is hard for me to see its purpose at the moment

File: packages/jira-adapter/src/client/client.ts (hunk 13 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:111]: Nice catch, thanks!

File: packages/jira-adapter/src/client/client.ts [hunk 14]
```diff
@@ -247,4 +350,59 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
...
     })
   }
+
+  public async getCloudId(): Promise<string> {
+    if (this.cloudId === undefined) {
+      const response = await this.get({
+        url: GET_CLOUD_ID_URL,
+      })
+      if (!isCloudIdResponse(response.data)) {
+        throw new Error(`'Failed to get cloud id, received invalid response' ${response.data}`)
```
Comment (shayc331) [inline packages/jira-adapter/src/client/client.ts:360]: again Im not sure if we want to throw here, it is more reasonable than the other throws (because cloudId is essential for the deployment) but if we call it from a change validator we dont want to throw, or we need to add try catch to not fail the deployment

File: packages/jira-adapter/test/change_validators/workflowsV2/missing_extensions_transition_rules.test.ts [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+const NON_EXISTENT_CONNECT_EXTENSION_ID = 'non-existent-extension'
+const NON_EXISTENT_CONNECT_EXTENSION_TRANSITION_RULE = createConnectTransitionRule(NON_EXISTENT_CONNECT_EXTENSION_ID)
+const NON_EXISTENT_FORGE_EXTENSION_ID = 's'.repeat(EXTENSION_ID_LENGTH)
+const NON_EXISTENT_FORGE_EXTENSION_TRANSITION_RULE = createForgeTransitionRule(NON_EXISTENT_FORGE_EXTENSION_ID)
+const NON_EXISTENT_EXTENSIONS_TRANSITION_RULES = [
+  NON_EXISTENT_CONNECT_EXTENSION_TRANSITION_RULE,
+  NON_EXISTENT_FORGE_EXTENSION_TRANSITION_RULE,
+]
+
+describe('missingAppsTransitionRulesReferencedWorkflowDeletionChangeValidator', () => {
```
Comment (shayc331) [inline packages/jira-adapter/test/change_validators/workflowsV2/missing_extensions_transition_rules.test.ts:61]: please add a test for a valid flow without an error

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+
+export const getRuleTypeFromWorkflowTransitionRule = (transitionRule: WorkflowV2TransitionRule): RuleType =>
+  getRuleTypeFromRuleKey(transitionRule.ruleKey)
+
+const getExtensionKeyFromWorkflowTransitionRule = (transitionRule: WorkflowV2TransitionRule): string | undefined => {
+  const ruleType = getRuleTypeFromWorkflowTransitionRule(transitionRule)
+
+  switch (ruleType) {
+    case RuleType.Connect:
+      return _.get(transitionRule, 'parameters.appKey', undefined)
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts:120]: if we have a dedicated type for each ruleType we won't need this _.get

File: packages/jira-adapter/src/filters/workflow/transition_structure.ts [hunk 17]
```diff
@@ -160,15 +160,15 @@ export const walkOverTransitionIds = (transition: WorkflowTransitionV1, func: (v
...
+        !_.isEmpty(_.get(action, 'parameters.scriptRunner.transitionId')),
     )
     .forEach(action => {
-      func(action.parameters.scriptRunner)
+      func(_.get(action, 'parameters.scriptRunner'))
```
Comment (shayc331) [inline packages/jira-adapter/src/filters/workflow/transition_structure.ts:171]: we can add it to the relevant ruleType type

File: packages/jira-adapter/src/client/client.ts [hunk 18]
```diff
@@ -41,6 +42,33 @@ const DEFAULT_PAGE_SIZE: Required<definitions.ClientPageSizeConfig> = {
 
 const RATE_LIMIT_HEADER_PREFIX = 'x-ratelimit-'
 
+export const GET_CLOUD_ID_URL = '/_edge/tenant_info'
+export const GQL_BASE_URL_GIRA = '/rest/gira/1'
+export const GQL_BASE_URL_GATEWAY = '/gateway/api/graphql'
+
```
Comment (shayc331) [inline packages/jira-adapter/src/client/client.ts:48]: @IdoZakk please go over the client code changes

**Review by shayc331 — COMMENTED** [review]
💣 looks really good, add some comments. 💣 
please reach out if you wanna discuss my thoughts.

File: packages/jira-adapter/src/change_validators/workflows/empty_validator_workflow.ts (hunk 3 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflows/empty_validator_workflow.ts:69]: I haven't seen it anywhere... I'm assuming it isn't a required part of system/connect/forge

File: packages/jira-adapter/src/client/client.ts [hunk 19]
```diff
@@ -58,9 +86,83 @@ const isGraphQLResponse = createSchemeGuard<graphQLResponseType>(
...
+    ecosystem: Joi.object({
+      appInstallationsByContext: Joi.object({
+        nodes: Joi.array()
+          .items({
+            app: Joi.object({
+              id: Joi.string().required(),
+              name: Joi.string().required(),
+            }).required(),
+          })
+          .required(),
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:150]: Will do.

File: packages/jira-adapter/src/client/client.ts [hunk 20]
```diff
@@ -247,4 +350,59 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
...
+  private async getInstalledExtensionsUPM(): Promise<ExtensionType[]> {
+    const upmResponse = await this.get({
+      url: '/rest/plugins/1.0/',
+    })
+    if (isUPMInstalledAppsResponse(upmResponse)) {
+      return upmResponse.data.plugins
+        .filter(plugin => plugin.enabled && plugin.userInstalled)
+        .map(plugin => ({ name: plugin.name, id: plugin.key }))
+    }
+    console.log(upmResponse)
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:381]: Yes :)

File: packages/jira-adapter/src/client/client.ts [hunk 21]
```diff
@@ -247,4 +350,59 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
...
+        const id = node.app.id.split('/')[1]
+        const { name } = node.app
+        return { id, name }
+      })
+    }
+
+    throw new Error('Failed to get GQL Gateway Installed Apps response')
+  }
+
+  public async getInstalledExtensions(): Promise<ExtensionType[]> {
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:405]: This function will be called in a fix element as well.

File: packages/jira-adapter/src/filters/workflowV2/types.ts (hunk 6 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/filters/workflowV2/types.ts:88]: It is a pretty loose scheme, specifically connect apps have parameters.appKey and forge apps have parameters.key

These are the only enforced fields except for ruleKey which they all have (parallel to parameters, not inside it)

File: packages/jira-adapter/test/change_validators/workflowsV2/missing_extensions_transition_rules.test.ts (hunk 15 above)
Comment (IlayDavid95) [reply packages/jira-adapter/test/change_validators/workflowsV2/missing_extensions_transition_rules.test.ts:61]: Nice catch, thanks!

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+  }
+}
+
+export const getInstalledExtensionsMap = async (client: JiraClient): Promise<Record<string, ExtensionType>> =>
+  awu(await client.getInstalledExtensions()).reduce(
+    (acc, app) => {
+      acc[app.id] = app
+      return acc
+    },
+    {} as Record<string, ExtensionType>,
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts:157]: Fixed

File: packages/jira-adapter/src/client/client.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:150]: I'm pretty sure because these are the only fields queried in the GQL request.

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts (hunk 16 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts:120]: Fixed, there is no need for this get anyway.

File: packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+      .filter(isWorkflowV2Instance)
+      .flatMap(getTransitionRulesWithElemID)
+      .filter(transitionRule => getRuleTypeFromWorkflowTransitionRule(transitionRule) !== RuleType.System)
+      .toArray()
+
+    // The following steps perform requests to Jira API and so we first make sure it is required.
+    if (_.isEmpty(relevantRules)) {
+      return []
+    }
+    const installedExtensionsMap = await getInstalledExtensionsMap(client)
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/workflowsV2/missing_extensions_transition_rules.ts:176]: fixed.

File: packages/jira-adapter/src/client/client.ts (hunk 14 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:360]: I think we should wrap with try/catch. the change validator doesn't use this anyway.

File: packages/jira-adapter/src/client/client.ts [hunk 24]
```diff
@@ -247,4 +350,59 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
...
+    const upmResponse = await this.get({
+      url: '/rest/plugins/1.0/',
+    })
+    if (isUPMInstalledAppsResponse(upmResponse)) {
+      return upmResponse.data.plugins
+        .filter(plugin => plugin.enabled && plugin.userInstalled)
+        .map(plugin => ({ name: plugin.name, id: plugin.key }))
+    }
+    console.log(upmResponse)
+    throw new Error('Failed to get UPM Installed Apps response')
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:382]: We throw in all the other places though.

File: packages/jira-adapter/src/client/client.ts [hunk 25]
```diff
@@ -247,4 +350,59 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
...
+    if (isGQLGatewayInstalledAppsResponse(gqlResponse)) {
+      return gqlResponse.data.ecosystem.appInstallationsByContext.nodes.map(node => {
+        // app.id is of the format: <prefix>/<app-id>
+        const id = node.app.id.split('/')[1]
+        const { name } = node.app
+        return { id, name }
+      })
+    }
+
+    throw new Error('Failed to get GQL Gateway Installed Apps response')
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:402]: It is wrapped in a try/catch in the cv.

File: packages/jira-adapter/src/filters/workflow/transition_structure.ts (hunk 5 above)
Comment (shayc331) [reply packages/jira-adapter/src/filters/workflow/transition_structure.ts:167]: please remove the _.get here

**Review by shayc331 — COMMENTED** [review]
good job!
Im waiting for @IdoZakk to review the client code. approving from my side

File: packages/jira-adapter/src/client/client.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (IdoZakk) [reply packages/jira-adapter/src/client/client.ts:404]: IIUC this is no longer the case, so please remove it from the client.
In general, the client class should be minimal and contain only functions that are relevant to all/some client calls.
The cloud id is a good candidate for the client, it is in the essence of calling outside REST APIs.
This function is used to store information on the client. It is true that the info is brought from the client, but this is not its purpose. We can store information in a few other locations if needed

**Review by IdoZakk — COMMENTED** [review]
Added my review on the client :-)

File: packages/jira-adapter/src/client/client.ts [hunk 26]
```diff
@@ -247,4 +350,58 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
       },
     })
   }
+
+  public async getCloudId(): Promise<string> {
+    if (this.cloudId === undefined) {
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:355]: I've done it in the refactor PR. Will cherry pick the change to here.

File: packages/jira-adapter/src/client/client.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:404]: @IdoZakk @shayc331

[the remark this answers is not in the document]
Comment (IdoZakk) [reply packages/jira-adapter/src/client/client.ts:404]: Should we keep them? aren't we using them just for the validator?

[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:404]: While we use it in the validator I'm not sure cohesion wise it is part of the validator module. This is a query of _installed apps_ in jira environment. It might be used in a change validator for _workflow transition rules_ but that doesn't mean it should be in that module. We have a fresh example in the form of getCloudId, just because it was only used in some filter in workflow doesn't mean it should be there.

File: packages/jira-adapter/src/client/client.ts [hunk 27]
```diff
@@ -247,4 +350,58 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
...
+      }
+      this.cloudId = response.data.cloudId
+    }
+
+    return this.cloudId
+  }
+
+  public setCloudId(cloudId?: string): void {
+    this.cloudId = cloudId
+  }
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:370]: removed

File: packages/jira-adapter/src/client/client.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:381]: handled with a try catch in the cv

File: packages/jira-adapter/src/common/workflow/transition_rules.ts [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+
+import { WorkflowV2TransitionRule } from '../../filters/workflowV2/types'
+
+export enum RuleType {
+  Connect = 'connect',
+  Forge = 'forge',
+  System = 'system',
+  Invalid = 'invalid',
+}
+export const isValidRuleType = (value: any): value is RuleType =>
```
Comment (shayc331) [inline packages/jira-adapter/src/common/workflow/transition_rules.ts:25]: nit: please remove unnecessary import

File: packages/jira-adapter/src/common/extensions.ts [hunk 29]
```diff
@@ -0,0 +1,161 @@
...
+      const { name } = node.app
+      return { id, name }
+    })
+  }
+
+  throw new Error('Failed to get GQL Gateway Installed Apps response')
+}
+
+export const getInstalledExtensions = async (client: JiraClient): Promise<ExtensionType[]> => [
+  ...(await getInstalledExtensionsUPM(client)),
```
Comment (shayc331) [inline packages/jira-adapter/src/common/extensions.ts]: nit: lets try to parallel this one, if it takes you more then 10 min leave it

**Review by shayc331 — APPROVED** [review]
looks very good, had some nits
approved from my side, please wait for @IdoZakk to approve the client side.
please make sanity checks for workflow before merging because we change the scheme and it may affect it, the WorkflowV2 is not a part of the e2e yet (fetch and deploy of workflow with exist or not exist extensions).

File: packages/jira-adapter/src/common/extensions.ts (hunk 29 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/common/extensions.ts]: done

File: packages/jira-adapter/src/common/workflow/transition_rules.ts (hunk 28 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/common/workflow/transition_rules.ts:25]: Done

File: packages/jira-adapter/src/client/client.ts [hunk 30]
```diff
@@ -72,6 +91,7 @@ export default class JiraClient extends clientUtils.AdapterHTTPClient<Credential
       timeout: DEFAULT_TIMEOUT_OPTS,
     })
     this.isDataCenter = clientOpts.isDataCenter
+    this.cloudId = undefined
```
Comment (IdoZakk) [inline packages/jira-adapter/src/client/client.ts:94]: Is this line needed?

**Review by IdoZakk — COMMENTED** [review]
The client looks good, look at my small comments

File: packages/jira-adapter/src/client/client.ts (hunk 30 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:94]: Removed.

File: packages/jira-adapter/src/client/client.ts [hunk 31]
```diff
@@ -41,6 +41,25 @@ const DEFAULT_PAGE_SIZE: Required<definitions.ClientPageSizeConfig> = {
...
+
+const CLOUD_ID_RESPONSE_SCHEME = Joi.object({
+  cloudId: Joi.required(),
+})
+  .unknown(true)
+  .required()
+
+const isCloudIdResponse = createSchemeGuard<CloudIdResponseType>(
+  CLOUD_ID_RESPONSE_SCHEME,
+  'Failed to get cloud id response',
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/client/client.ts:60]: Changed.

**Review by IdoZakk — APPROVED** [review]
🚀 👨‍🎤
